ckanext/csrf_filter/anti_csrf.py

```diff
@@ -367,7 +367,13 @@ def apply_token(response, request=None):
     """ Rewrite HTML to insert tokens if applicable.
     If a new token is generated, it will be added to 'response' as a cookie.
     """
-    html = getattr(response, 'data', None)
+    try:
+        # If the response data can't be decoded from bytes to str
+        # we can't insert a token
+        html = response.get_data(as_text=True)
+    except UnicodeDecodeError:
+        html = None
```

Comment:
I ended up trying this approach after getting screeds of UnicodeDecodeErrors in the logs, all related to being unable to decode the `bytes` of a responses data to `str`. See the commit message:
```
    apply_token is called from a Flask after_app_request
    hook, which fires for any response handled by Flask
    including assets like images/css/fonts. This will
    hopefully prevent it failing on data that can't be
    parsed for possible token insertion.
```
I wonder if the after_app_request hook could do some filtering, maybe on `response.mimetype` or something like that?

Reply:
will this work with 2.8/2.7?

Reply:
This will only get called in CKAN 2.8 or greater, as 2.7 was Pylons only which seems to use the `anti_csrf_pylons.py` version instead. The werkzeug version in CKAN 2.8.8 is 0.14.1, which has the same get_data implementation.

https://github.com/pallets/werkzeug/blob/0.14.1/werkzeug/wrappers.py#L973

Looks like it should be fine for CKAN 2.8 and 2.9 :+1:

Reply:
How feasible is it to create a test case for a response payload that can't be decoded?